pkg/util/buffered_chan.go

```diff
@@ -0,0 +1,94 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+package util
+
+import (
+	"context"
+	"sync"
+)
+
+// BufferedChan behaves like a `chan []interface{}` (See thread safety for restrictions), but is
+// most efficient as it uses internally a channel of []interface{}. This reduces the number of reads
+// and writes to the channel. Instead of having one write and one read for each value for a regular channel,
+// there are one write and one read for each `bufferSize` value.
+// Thread safety:
+// 	- `BufferedChan.Put` cannot be called concurrently.
+// 	- `BufferedChan.Get` cannot be called concurrently.
+//  - `BufferedChan.Put` can be called while another goroutine calls `BufferedChan.Get`.
+type BufferedChan struct {
+	c        chan []interface{}
+	pool     *sync.Pool
+	putSlice []interface{}
+	getSlice []interface{}
+	getIndex int
+	ctx      context.Context
+}
+
+// NewBufferedChan creates a new instance of `BufferedChan`.
+// `ctx` can be used to cancel all Put and Get operations.
+func NewBufferedChan(ctx context.Context, chanSize int, bufferSize int) *BufferedChan {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return make([]interface{}, 0, bufferSize)
+		},
+	}
+	return &BufferedChan{
+		c:        make(chan []interface{}, chanSize),
+		pool:     pool,
+		putSlice: pool.Get().([]interface{}),
+		ctx:      ctx,
+	}
+}
+
+// Put puts a new value into c.
+// Cannot be called concurrently.
+func (c *BufferedChan) Put(value interface{}) {
+	if cap(c.putSlice) <= len(c.putSlice) {
+		select {
+		case c.c <- c.putSlice:
+		case <-c.ctx.Done():
```

Comment:
Shouldn't put do something if the operation failed? Return an error or panic?

Reply:
Which operation are you talking about, `c.c <- c.putSlice`  or `<-c.ctx.Done()`?

Reply:
I mean when `<-c.ctx.Done()` branch is taken, it would mean we failed to write into the channel and dropped a batch of series. I don't think we want this to happen silently.